compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir

```diff
@@ -0,0 +1,163 @@
+
+hal.executable private @matmul_static_dispatch_0 {
+  hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
+
+    hal.executable.export public @matmul_static_dispatch_0_matmul_1024x4096x12345 ordinal(0) layout(#pipeline_layout) {
+    // Check that num_threads is reflected in the workgroup size.
+    // CHECK-LABEL: hal.executable.export public @matmul_static_dispatch_0_matmul_1024x4096x12345
+    // CHECK-DAG: %[[c32:.*]] = arith.constant 32 : index
+    // CHECK-DAG: %[[c1:.*]] = arith.constant 1 : index
+    // CHECK-DAG: %[[c1_0:.*]] = arith.constant 1 : index
+    // CHECK: hal.return %[[c32]], %[[c1]], %[[c1_0]] : index, index, index
+    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index):
+      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2, %arg3
+      hal.return %x, %y, %z : index, index, index
+    }
+
+    builtin.module {
+      func.func @matmul_static_dispatch_0_matmul_1024x4096x12345() {
+        // Check that the tiling matches num_threads.
+        // CHECK-LABEL: func.func @matmul_static_dispatch_0_matmul_1024x4096x12345
+        %c0 = arith.constant 0 : index
+        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<1024x12345xf32>>
+        %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<12345x4096xf32>>
+        %2 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<readwrite:tensor<1024x4096xf32>>
+        %3 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [1024, 12345], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<1024x12345xf32>> -> tensor<1024x12345xf32>
+        %4 = flow.dispatch.tensor.load %1, offsets = [0, 0], sizes = [12345, 4096], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<12345x4096xf32>> -> tensor<12345x4096xf32>
+        %5 = flow.dispatch.tensor.load %2, offsets = [0, 0], sizes = [1024, 4096], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<1024x4096xf32>> -> tensor<1024x4096xf32>
+        %6 = linalg.matmul ins(%3, %4 : tensor<1024x12345xf32>, tensor<12345x4096xf32>) outs(%5 : tensor<1024x4096xf32>) -> tensor<1024x4096xf32>
+        flow.dispatch.tensor.store %6, %2, offsets = [0, 0], sizes = [1024, 4096], strides = [1, 1] : tensor<1024x4096xf32> -> !flow.dispatch.tensor<readwrite:tensor<1024x4096xf32>>
+        return
+      }
+    }
+  }
+}
+
+transform.sequence failures(propagate) {
+^bb1(%variant_op: !pdl.operation):
+  %original_matmul = transform.structured.match ops{["linalg.matmul"]} in %variant_op
+    : (!pdl.operation) -> !pdl.operation
+
+  transform.iree.compute_workgroup_count_region %original_matmul
```

Comment:
I am not sure this is always going to work.... you also need to ensure that the corresponding tiling is also consistent with this (AFAIK based on how the `scf.for_all` is normalized

Reply:
Yes, I think I needed a test like

```
%0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
%1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
%4 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
transform.iree.compute_workgroup_count_region %2 num_threads [256]
```

Our use case was to do something like below test case in IREE, but we couldn't do hence we pushed a [patch](https://reviews.llvm.org/D147830#change-Z5DR4wmNOs1v)%5D) in MLIR to support the same and it needs this patch, hence we tried to upstream it, this patch just extracts the logic of workgroup count. 

Does your patch handle the use case we are trying to achieve with coalesce?
```

%0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
%1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
%3 = transform.loop.coalesce_parallel %1 { collapsed_dim0 = array<i64: 0, 1>, mapping = [#gpu.block<x>] } : (!pdl.operation) -> !pdl.operation
%4 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
```

Thanks again for reviewing.

Reply:
> Yes, I think I needed a test like
> 
> ```
> %0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> %1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
> transform.iree.compute_workgroup_count_region %2 num_threads [256]
> ```
> 
> Our use case was to do something like below test case in IREE, but we couldn't do hence we pushed a [patch](https://reviews.llvm.org/D147830#change-Z5DR4wmNOs1v)%5D) in MLIR to support the same and it needs this patch, hence we tried to upstream it, this patch just extracts the logic of workgroup count.
> 
> Does your patch handle the use case we are trying to achieve with coalesce?
> 
> ```
> 
> %0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> %1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
> %3 = transform.loop.coalesce_parallel %1 { collapsed_dim0 = array<i64: 0, 1>, mapping = [#gpu.block<x>] } : (!pdl.operation) -> !pdl.operation
> %4 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> ```
> 
> Thanks again for reviewing.

I think the coalescing should be folded into the `structured.tile_to_forall_op` (thats how I think it is handled today in IREE).
If you can post the IR after `transform.loop.coalese_parallel` I can let you know if it works or not. I suspect it does.